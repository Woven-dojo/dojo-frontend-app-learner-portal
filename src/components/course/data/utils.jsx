import {
  COURSE_AVAILABILITY_MAP,
  COURSE_MODES_MAP,
  COURSE_PACING_MAP,
  LICENSE_SUBSIDY_TYPE,
  OFFER_SUBSIDY_TYPE,
} from './constants';

export function hasCourseStarted(start) {
  const today = new Date();
  const startDate = new Date(start);
  return startDate && today >= startDate;
}

export function findUserEnrollmentForCourseRun({ userEnrollments, key }) {
  return userEnrollments.find(
    ({
      isEnrollmentActive,
      isRevoked,
      courseRunId,
    }) => (isEnrollmentActive && !isRevoked && courseRunId === key),
  );
}

export function isUserEntitledForCourse({ userEntitlements, courseUuid }) {
  return userEntitlements.some(({ courseUuid: uuid }) => uuid === courseUuid);
}

export function weeksRemainingUntilEnd(courseRun) {
  const today = new Date();
  const end = new Date(courseRun.end);
  const secondsDifference = Math.abs(end - today) / 1000;
  const days = Math.floor(secondsDifference / 86400);
  return Math.floor(days / 7);
}

export function hasTimeToComplete(courseRun) {
  return courseRun.weeksToComplete <= weeksRemainingUntilEnd(courseRun);
}

export function isArchived(courseRun) {
  if (courseRun.availability) {
    return courseRun.availability === COURSE_AVAILABILITY_MAP.ARCHIVED;
  }
  return false;
}

export function isCourseSelfPaced(pacingType) {
  return pacingType === COURSE_PACING_MAP.SELF_PACED;
}

export function isCourseInstructorPaced(pacingType) {
  return pacingType === COURSE_PACING_MAP.INSTRUCTOR_PACED;
}

export function createCourseInfoUrl({ baseUrl, courseKey }) {
  return `${baseUrl}/courses/${courseKey}/info`;
}

export const numberWithPrecision = (number, precision = 2) => number.toFixed(precision);

// See https://openedx.atlassian.net/wiki/spaces/WS/pages/1045200922/Enroll+button+and+Course+Run+Selector+Logic
// for more detailed documentation on course run selection and the enroll button.
export function getActiveCourseRun(course) {
  return course.courseRuns.find(courseRun => courseRun.uuid === course.advertisedCourseRunUuid);
}

export function getAvailableCourseRuns(course) {
  return course.courseRuns.filter(
    courseRun => courseRun.isMarketable && courseRun.isEnrollable && !isArchived(courseRun),
  );
}

export function findOfferForCourse(offers, catalogList = []) {
  const offerIndex = offers.findIndex((offer) => catalogList?.includes(offer.catalog));
  if (offerIndex !== -1) {
    return offers[offerIndex];
  }
  return null;
}

const getBestCourseMode = (courseModes) => {
  const {
    VERIFIED, PROFESSIONAL, NO_ID_PROFESSIONAL, AUDIT, HONOR,
  } = COURSE_MODES_MAP;
  /** Returns the 'highest' course mode available.
    *  Modes are ranked ['verified', 'professional', 'no-id-professional', 'audit', 'honor'] */
  if (courseModes.includes(VERIFIED)) {
    return VERIFIED;
  }
  if (courseModes.includes(PROFESSIONAL)) {
    return PROFESSIONAL;
  }
  if (courseModes.includes(NO_ID_PROFESSIONAL)) {
    return NO_ID_PROFESSIONAL;
  }
  if (courseModes.includes(AUDIT)) {
    return AUDIT;
  }
  if (courseModes.includes(HONOR)) {
    return HONOR;
  }
  return null;
};

export function findHighestLevelSeatSku(seats) {
  /** Returns the first seat found from the preferred course mode */
  if (!seats || seats.length <= 0) {
    return null;
  }
  const courseModes = seats.map(seat => seat.type);
  const courseMode = getBestCourseMode(courseModes);
  return seats.find(seat => seat.type === courseMode)?.sku;
}

export function shouldUpgradeUserEnrollment({
  userEnrollment,
  subscriptionLicense,
  enrollmentUrl,
}) {
  const isAuditEnrollment = userEnrollment?.mode === COURSE_MODES_MAP.AUDIT;
  if (isAuditEnrollment && subscriptionLicense && enrollmentUrl) {
    return true;
  }
  return false;
}

export function hasLicenseSubsidy(subsidy) {
  return subsidy?.subsidyType === LICENSE_SUBSIDY_TYPE;
}

export function hasOfferSubsidy(subsidy) {
  return subsidy?.subsidyType === OFFER_SUBSIDY_TYPE;
}

// Truncate a string to less than the maxLength characters without cutting the last word and append suffix at the end
export function shortenString(str, maxLength, suffix, separator = ' ') {
  if (str.length <= maxLength) { return str; }
  return `${str.substr(0, str.lastIndexOf(separator, maxLength))}${suffix}`;
}
