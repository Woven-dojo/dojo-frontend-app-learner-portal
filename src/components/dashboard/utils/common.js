export const setDashIfEmpty = (item, key, callback) => {
  if (!(key in item) || !item[key]) {
    return '-';
  }
  return callback(item[key]);
};

export const isElementInDOM = (step) => {
  const nodes = document.querySelectorAll(step.selector);
  return nodes.length > 0;
};
