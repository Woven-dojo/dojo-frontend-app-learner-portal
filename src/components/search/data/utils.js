// eslint-disable-next-line import/prefer-default-export
export const sortItemsByLabelAsc = items => (
  items.sort((a, b) => a.label.localeCompare(b.label))
);
