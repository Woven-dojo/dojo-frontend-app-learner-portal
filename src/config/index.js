import { hasFeatureFlagEnabled } from '@edx/frontend-enterprise-utils';

import {
  FEATURE_ENROLL_WITH_CODES,
  FEATURE_ENABLE_PATHWAYS,
  FEATURE_ENABLE_AUTO_APPLIED_LICENSES,
  FEATURE_BROWSE_AND_REQUEST,
} from './constants';

const features = {
  ENABLE_AUTO_APPLIED_LICENSES: (
    process.env.ENABLE_AUTO_APPLIED_LICENSES || hasFeatureFlagEnabled(FEATURE_ENABLE_AUTO_APPLIED_LICENSES)
  ),
  ENROLL_WITH_CODES: process.env.FEATURE_ENROLL_WITH_CODES || hasFeatureFlagEnabled(FEATURE_ENROLL_WITH_CODES),
  ENABLE_PATHWAYS: process.env.FEATURE_ENABLE_PATHWAYS || hasFeatureFlagEnabled(FEATURE_ENABLE_PATHWAYS),
  FEATURE_BROWSE_AND_REQUEST: process.env.FEATURE_BROWSE_AND_REQUEST
    || hasFeatureFlagEnabled(FEATURE_BROWSE_AND_REQUEST),
};

export { features };
