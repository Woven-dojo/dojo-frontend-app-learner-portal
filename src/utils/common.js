import moment from 'moment';

const courseEnded = (endDate) => moment(endDate) < moment();

export {
  // eslint-disable-next-line import/prefer-default-export
  courseEnded,
};
