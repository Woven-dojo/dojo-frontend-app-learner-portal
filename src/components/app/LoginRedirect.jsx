import PropTypes from 'prop-types';
import { useParams } from 'react-router-dom';
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
import { getLoginUrl } from '../utils/urls';

/**
 * This wrapper component redirects the user to the enterprise proxy login view with additional query
 * parameters if they are not already authenticated. If the user is already authenticated, simply return
 * the children to render the rest of the application.
 *
 * @param {node} children The child nodes to render if there is an authenticated user.
 * @param {element} loadingDisplay A React element to display while authenticated user is loading
 */

export default function LoginRedirect({
  children,
  loadingDisplay: LoadingDisplay,
}) {
  const user = getAuthenticatedUser();

  if (user) {
    return children;
  }

  const {
    enterpriseCustomerInviteKey,
  } = useParams();
  const enterpriseSlug = window.location.host.split('.')[0];
  global.location.href = getLoginUrl(enterpriseSlug, enterpriseCustomerInviteKey);
  return LoadingDisplay;
}
LoginRedirect.propTypes = {
  children: PropTypes.node,
  loadingDisplay: PropTypes.element,
};
LoginRedirect.defaultProps = {
  children: null,
  loadingDisplay: null,
};
