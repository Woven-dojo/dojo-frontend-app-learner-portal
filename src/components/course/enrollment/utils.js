/* eslint-disable import/prefer-default-export */
import { enrollButtonTypes } from '../data/constants';

const {
  ENROLL_DISABLED,
  TO_COURSEWARE_PAGE,
  TO_DATASHARING_CONSENT,
  TO_ECOM_BASKET,
  TO_VOUCHER_REDEEM,
  VIEW_ON_DASHBOARD,
} = enrollButtonTypes;

/**
 * The main logic to determine enrollment type (scenario).
 */
export function determineEnrollmentType({
  subsidyData: {
    subscriptionLicense, userSubsidy, enrollmentUrl, courseHasOffer,
  },
  isUserEnrolled,
  isEnrollable,
  isCourseStarted,
}) {
  if (isUserEnrolled) {
    return isCourseStarted ? TO_COURSEWARE_PAGE : VIEW_ON_DASHBOARD;
  }
  if (!isEnrollable) { return ENROLL_DISABLED; }
  if (!enrollmentUrl) { return ENROLL_DISABLED; }
  if (subscriptionLicense && userSubsidy) { return TO_DATASHARING_CONSENT; }
  if (subscriptionLicense && !userSubsidy) { return TO_ECOM_BASKET; }
  if (!subscriptionLicense && !courseHasOffer) { return TO_ECOM_BASKET; }
  if (!subscriptionLicense && courseHasOffer) { return TO_VOUCHER_REDEEM; }
  return ENROLL_DISABLED;
}
