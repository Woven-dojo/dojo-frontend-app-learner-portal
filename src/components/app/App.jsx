import React, { useEffect } from 'react';
import { Switch } from 'react-router-dom';
import useHotjar from 'react-use-hotjar';
import { AppProvider, AuthenticatedPageRoute, PageRoute } from '@edx/frontend-platform/react';

import NotFoundPage from '../NotFoundPage';
import NoticesProvider from '../notices-provider';
import {
  EnterpriseCustomerRedirect,
  EnterprisePageRedirect,
} from '../enterprise-redirects';
import { DashboardPage } from '../dashboard';
import { CoursePage } from '../course';
import { ProgramPage } from '../program';
import { SentryErrorBoundary } from '../sentry';
import ProgramProgressPage from '../program-progress/ProgramProgressPage';
import { SearchPage } from '../search';
import { LicenseActivationPage } from '../license-activation';
import { SkillsQuizPage } from '../skills-quiz';
import { EnterpriseInvitePage } from '../enterprise-invite';
import { ErrorPage } from '../error-page';

import { features } from '../../config';
import { ToastsProvider, Toasts } from '../Toasts';

export default function App() {
  if (process.env.HOTJAR_APP_ID) {
    const { initHotjar } = useHotjar();
    useEffect(() => {
      initHotjar(process.env.HOTJAR_APP_ID, process.env.HOTJAR_VERSION, process.env.HOTJAR_DEBUG);
    }, [initHotjar]);
  }

  const content = <>
    <Toasts />
    <Switch>
      <AuthenticatedPageRoute exact path="/" component={EnterpriseCustomerRedirect} />
      <AuthenticatedPageRoute exact path="/r/:redirectPath+" component={EnterprisePageRedirect} />
      <PageRoute exact path="/invite/:enterpriseCustomerInviteKey" component={EnterpriseInvitePage} />
      <PageRoute exact path="/:enterpriseSlug" component={DashboardPage} />
      <PageRoute exact path="/:enterpriseSlug/search" component={SearchPage} />
      <PageRoute exact path="/:enterpriseSlug/course/:courseKey" component={CoursePage} />
      {features.ENABLE_PROGRAMS && (
        <PageRoute exact path="/:enterpriseSlug/program/:programUuid" component={ProgramPage} />
      )}
      <PageRoute exact path="/:enterpriseSlug/program-progress/:programUUID" component={ProgramProgressPage} />
      <PageRoute exact path="/:enterpriseSlug/licenses/:activationKey/activate" component={LicenseActivationPage} />
      <PageRoute exact path="/:enterpriseSlug/skills-quiz" component={SkillsQuizPage} />
      <PageRoute path="*" component={NotFoundPage} />
    </Switch>
  </>

  return (
    <AppProvider>
      <NoticesProvider>
        <ToastsProvider>
          <SentryErrorBoundary renderError={renderError}>
            {content}
          </SentryErrorBoundary>
        </ToastsProvider>
      </NoticesProvider>
    </AppProvider>
  );
}

const renderError = (error) => (
  <ErrorPage message={error.message} />
)
