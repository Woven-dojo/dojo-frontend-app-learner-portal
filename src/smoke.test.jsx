import React from 'react';
import AxiosMockAdapter from 'axios-mock-adapter';
import renderer from 'react-test-renderer';
import axios from 'axios';
import { mergeConfig, getConfig, setConfig } from '@edx/frontend-platform/config';
import { initializeMockApp, history } from '@edx/frontend-platform';
import { App } from './components/app';
import { delay } from './utils/common';

import { createCourseEnrollmentWithStatus, COURSE_STATUSES } from './utils/tests/factories/enrollment';

// For modals
jest.mock('react-focus-on', () => ({
  FocusOn: (props) => {
    // eslint-disable-next-line
    const { children, ...otherProps } = props;
    return <focus-on {...otherProps}>{children}</focus-on>;
  },
}));

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: jest.fn((element) => <portal>{element}</portal>),
}));

jest.unmock('@edx/frontend-platform/logging');
jest.unmock('@edx/frontend-platform/analytics');
jest.unmock('@edx/frontend-platform/auth');

const mergeTestConfig = () =>
  mergeConfig({
    USE_API_CACHE: process.env.USE_API_CACHE || null,
    INTEGRATION_WARNING_DISMISSED_COOKIE_NAME: process.env.INTEGRATION_WARNING_DISMISSED_COOKIE_NAME || null,
    ENABLE_NOTICES: null,
    LEARNER_SUPPORT_URL: process.env.LEARNER_SUPPORT_URL || null,
    SENTRY_DSN: process.env.SENTRY_DSN || null,
    SENTRY_ENVIRONMENT: process.env.SENTRY_ENVIRONMENT || '',
    LOGIN_URL: `${process.env.LMS_BASE_URL}/login`,
    SENTRY_PROJECT_ENV_PREFIX: 'dojo-frontend-app-learner-portal',
    ACCESS_TOKEN_COOKIE_NAME: 'dojo-access-token',
    CREDENTIALS_BASE_URL: 'http://example-cred.com',
    CSRF_TOKEN_API_PATH: 'http://exaple-csrf-token-api.com',
    PUBLISHER_BASE_URL: 'http://example-publisher-url.com',
    IGNORED_ERROR_REGEX: '',
    LANGUAGE_PREFERENCE_COOKIE_NAME: 'dojo-language',
    STUDIO_BASE_URL: process.env.STUDIO_BASE_URL || null,
    ORDER_HISTORY_URL: 'http://example-order-history.com',
    REFRESH_ACCESS_TOKEN_ENDPOINT: '',
  });

const axiosMock = new AxiosMockAdapter(axios);
const { LMS_BASE_URL } = process.env;

const ENTERPRISE_SLUG = 'test-enterprise-id';
const ENTERPRISE_UUID = '55b3a044-c9f9-4550-9ce4-3ab366b2e9a6';
const ENTERPRISE_CUSTOMER_UUID = ENTERPRISE_UUID;
const ENTERPRISE_CATALOG_UUID = 'f62ce957-8f5a-452d-bc79-b5ef2651aab8';

const ENTERPRISE = {
  uuid: ENTERPRISE_UUID,
  name: 'Test Enterprise',
  slug: ENTERPRISE_SLUG,
  active: true,
  site: { domain: 'example.com', name: 'example.com' },
  enable_data_sharing_consent: true,
  enforce_data_sharing_consent: 'at_enrollment',
  branding_configuration: {
    enterprise_customer: ENTERPRISE_CUSTOMER_UUID,
    enterprise_slug: ENTERPRISE_SLUG,
    logo: 'http://localhost:18000/static/dojo-theme/images/logo.png',
    primary_color: '#00262B',
    secondary_color: '#EFF8FA',
    tertiary_color: '#0A7DA3',
  },
  enable_audit_enrollment: false,
  replace_sensitive_sso_username: false,
  enable_portal_code_management_screen: true,
  sync_learner_profile_data: false,
  enable_audit_data_reporting: false,
  enable_learner_portal: true,
  enable_portal_reporting_config_screen: true,
  enable_portal_saml_configuration_screen: true,
  enable_portal_subscription_management_screen: true,
  hide_course_original_price: false,
  enable_analytics_screen: false,
  enable_integrated_customer_learner_portal_search: true,
  enable_portal_lms_configurations_screen: true,
  sender_alias: null,
  identity_providers: [],
  enterprise_customer_catalogs: [ENTERPRISE_CATALOG_UUID],
  reply_to: null,
  enterprise_notification_banner: { title: '', text: '' },
  identity_provider: 'saml-test',
  hide_labor_market_data: false,
  contact_email: 'edx@example.com',
};

const ENTERPRISE_CUSTOMER_REPLY = {
  next: null,
  previous: null,
  count: 1,
  num_pages: 1,
  current_page: 1,
  start: 0,
  results: [ENTERPRISE],
};
axiosMock
  .onGet(new RegExp(`${LMS_BASE_URL}/enterprise/api/v1/enterprise-customer/*`))
  .reply(200, ENTERPRISE_CUSTOMER_REPLY);

const COURSE_ENROLLMENTS = [
  createCourseEnrollmentWithStatus(COURSE_STATUSES.inProgress),
  createCourseEnrollmentWithStatus(COURSE_STATUSES.completed),
  createCourseEnrollmentWithStatus(COURSE_STATUSES.upcoming),
];
const COURSE_ENROLLMENTS_REPLY = COURSE_ENROLLMENTS;
axiosMock
  .onGet(new RegExp(`${LMS_BASE_URL}/enterprise_learner_portal/api/v1/enterprise_course_enrollments/*`))
  .reply(200, COURSE_ENROLLMENTS_REPLY);

const LOADING_DELAY = 100;

const generateProfileImage = () => {
  const profileImage = { hasImage: false };
  const sizes = {
    largeUrlFull: 500,
    largeUrlLarge: 100,
    largeUrlMedium: 50,
    largeUrlSmall: 30,
  };

  Object.entries(sizes).forEach(([field, size]) => {
    profileImage[field] = `${LMS_BASE_URL}/static/images/profile/default_${size}.png`;
  });

  return profileImage;
};

const ENTERPRISE_LEARNER = {
  username: 'learner',
  name: 'john learner',
  roles: [],
  profileImage: generateProfileImage(),
};

jest.mock('./colors.scss', () => ({
  dark: 'grey',
  white: 'white',
  primary: 'blue',
  info100: 'blue',
  info500: 'blue',
}));

describe('[Smoke] Smoke test', () => {
  let prevConfig = {};

  afterAll(() => {
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    prevConfig = getConfig();
    mergeTestConfig();
  });

  afterEach(() => {
    setConfig(prevConfig);
  });

  it('Dashboard renders', async () => {
    history.push({ pathname: ENTERPRISE_SLUG });
    initializeMockApp({
      messages: [],
      authenticatedUser: ENTERPRISE_LEARNER,
    });

    await renderer.act(async () => {
      const tree = await renderer.create(<App />);
      await delay(LOADING_DELAY);

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
