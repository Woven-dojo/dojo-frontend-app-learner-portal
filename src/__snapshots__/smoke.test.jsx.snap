// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Smoke] Smoke test Dashboard renders 1`] = `
Array [
  <div
    className="position-absolute"
  >
    <a
      className="skip-nav-link sr-only sr-only-focusable btn btn-primary mt-3 ml-2"
      href="/test-enterprise-id#content"
      onClick={[Function]}
    >
      Skip to main content
    </a>
  </div>,
  <main
    id="content"
  >
    <div
      className="enterprise-banner bg-brand-secondary"
    >
      <div
        className="container-mw-lg container-fluid"
      >
        <div
          className="row banner-content"
        >
          <h1
            className="h2 mb-0 py-3 pl-3 border-brand-tertiary text-brand-secondary"
          >
            Test Enterprise
          </h1>
        </div>
      </div>
    </div>
    <div
      className="py-5 container-fluid"
    >
      <div
        className="loading-spinner d-flex justify-content-center align-items-center"
      >
        <div
          className="spinner-border text-primary"
          role="status"
        >
          <span
            className="sr-only"
          >
            loading your edX benefits from your organization
          </span>
        </div>
      </div>
    </div>
  </main>,
  <footer
    aria-label="edX Home"
    className="footer d-flex border-top py-3 px-4"
    role="contentinfo"
  >
    <div
      className="container-fluid d-flex"
    >
      <a
        aria-label="edX Home"
        className="d-block"
        href="https://open.edx.org"
      >
        <img
          alt="Powered by Open edX"
          style={
            Object {
              "maxHeight": 45,
            }
          }
        />
      </a>
      <div
        className="flex-grow-1"
      />
    </div>
  </footer>,
]
`;
